### PR TITLE
KAFKA-9712: Catch and handle exception thrown by reflections scanner

### PR DESCRIPTION
**Pull request by ncliang**
This commit works around a bug in v0.9.12 in upstream `reflections` library by catching and handling the exception thrown.

The reflections issue is tracked by:
https://github.com/ronmamo/reflections/issues/273

New unit tests were introduced to test the behavior.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by ncliang** [comment]
@rhauch @kkonstantine please take a look. The regression was introduced in 2.5, so this fix should also go back as far.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java [hunk 1]
```diff
@@ -339,9 +339,16 @@ private PluginScanResult scanPluginPath(
...
             ClassLoader loader
     ) throws InstantiationException, IllegalAccessException {
-        Set<Class<? extends T>> plugins = reflections.getSubTypesOf(klass);
-
         Collection<PluginDesc<T>> result = new ArrayList<>();
+        Set<Class<? extends T>> plugins;
+        try {
+             plugins = reflections.getSubTypesOf(klass);
+        } catch (ReflectionsException e) {
+            log.debug("Reflection scanner could not find any classes for URLs {}",
```
Comment (kkonstantine) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java]: nit: plural (`Reflections`) seems more appropriate because it refers to the library/class.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java [hunk 2]
```diff
@@ -339,9 +339,16 @@ private PluginScanResult scanPluginPath(
...
+                reflections.getConfiguration().getUrls());
+            return result;
```
Comment (kkonstantine) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java]: `result` is unused in this code block. To be future proof, I'd suggest being explicit by returning an empty list here, and declare `result` right above the block that is being used at.

Comment (kkonstantine) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java]:
```suggestion
            return Collections.emptyList();
```

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoaderTest.java [hunk 3]
```diff
@@ -17,15 +17,22 @@
...
 import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertTrue;
 
 public class DelegatingClassLoaderTest {
 
+    @Rule
+    public TemporaryFolder pluginPath = new TemporaryFolder();
+
+
```
Comment (kkonstantine) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoaderTest.java:37]: nit: extra blank line

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoaderTest.java [hunk 4]
```diff
@@ -61,4 +68,23 @@ public void testLoadingPluginClass() throws ClassNotFoundException {
             assertNotNull(classLoader.pluginClassLoader(pluginClassName));
         }
     }
+
```
Comment (kkonstantine) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoaderTest.java:73]: Nice to have test cases. Are all the cases that we want to cover? Did also fail with an empty directory, or a directory that contained an unparsable file (not a `jar`, but let's say a text file). If so, it'd be nice to have tests for these edge cases too. In general, as inspiration we could use how `/usr/share/java` looks on typical linux installations.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java [hunk 5]
```diff
@@ -339,9 +339,16 @@ private PluginScanResult scanPluginPath(
             Class<T> klass,
             ClassLoader loader
     ) throws InstantiationException, IllegalAccessException {
-        Set<Class<? extends T>> plugins = reflections.getSubTypesOf(klass);
-
         Collection<PluginDesc<T>> result = new ArrayList<>();
+        Set<Class<? extends T>> plugins;
+        try {
+             plugins = reflections.getSubTypesOf(klass);
```
Comment (kkonstantine) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java]: what happens if there's a mixture of valid `jars` and unrelated files in the path that is scanned. Will we miss plugins residing in that path, by skipping early here?

**Review by kkonstantine — COMMENTED** [review]
Thanks for the PR @ncliang 
In principle, I agree with the approach, but I have a few questions/suggestions.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java (hunk 5 above)
Comment (ncliang) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java]: In my tests, this scenario did not trigger the bug. It would only trigger when none of the files scanned contained a class. I can add a testcase for this.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java (hunk 2 above)
Comment (ncliang) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java]: I was unsure if the returned Collection would be mutated somewhere and was leaning on the safe side to return mutable `ArrayList` matching previous behavior. It does seem like they are not mutated though and an immutable empty list should be fine.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoaderTest.java (hunk 4 above)
Comment (ncliang) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoaderTest.java:73]: I have added more testcases to cover these scenarios.

**Review by ncliang — COMMENTED** [review]
Thanks for review @kkonstantine . I think I have addressed all your comments. Please take another look?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java (hunk 2 above)
Comment (kkonstantine) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java]: That's a good point too. But what I wanted to highlight is to be explicit and return the exact collection, that being `Collections.emptyList()` or `new ArrayList()` (the former should be fine as you noted), instead of returning what's stored in `result` (whose declaration is good to be close to the use as much as possible). 

That's to guard against `result` being used earlier by code in the future. Improbable, but also doesn't hurt and it's a good practice IMO.

**Comment by kkonstantine** [comment]
ok to test

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoaderTest.java [hunk 6]
```diff
@@ -61,4 +69,59 @@ public void testLoadingPluginClass() throws ClassNotFoundException {
...
+    @Test
+    public void testLoadingMixOfValidAndInvalidPlugins() throws Exception {
+        TestPlugins.assertAvailable();
+
+        pluginPath.newFile("invalid.jar");
+        pluginPath.newFolder("my-plugin");
+        pluginPath.newFile("my-plugin/invalid.jar");
+        for (String fullPath : TestPlugins.pluginPath()) {
+            FileUtils.copyFileToDirectory(new File(fullPath), pluginPath.getRoot());
```
Comment (kkonstantine) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoaderTest.java]: Great tests. 
I'd suggest using nio (`Files.copy` and `Paths`), rather than bringing in the dependency in the `org.codehaus.plexus` util, that is available now, but is not listed explicitly in the project dependencies. For simple things like these it's almost always better to use standard java library stuff in tests.

Comment (ncliang) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoaderTest.java]: Good point. And, in fact, the import is disallowed. noice.
```
DelegatingClassLoaderTest.java:20:1: Disallowed import - org.codehaus.plexus.util.FileUtils. [ImportControl]
```

**Comment by kkonstantine** [comment]
ok to test

**Comment by kkonstantine** [comment]
retest this please

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java [hunk 7]
```diff
@@ -339,8 +340,15 @@ private PluginScanResult scanPluginPath(
             Class<T> klass,
             ClassLoader loader
     ) throws InstantiationException, IllegalAccessException {
-        Set<Class<? extends T>> plugins = reflections.getSubTypesOf(klass);
-
+        Set<Class<? extends T>> plugins;
+        try {
+             plugins = reflections.getSubTypesOf(klass);
+        } catch (ReflectionsException e) {
+            log.debug("Reflections scanner could not find any classes for URLs {}",
```
Comment (chia7712) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java]: Could it print the origin exception also? ```log.debug("xx", e)```

Comment (ncliang) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java]: Good idea. Done.

**Comment by kkonstantine** [comment]
retest this please.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java (hunk 5 above)
Comment (kkonstantine) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java]: @ncliang checkstyle fails because there's an extra single white space here:

```
[ant:checkstyle] [ERROR] /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk11-scala2.13/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java:345: 'try' child has incorrect indentation level 13, expected level should be 12. [Indentation]
```

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java [hunk 8]
```diff
@@ -339,8 +340,15 @@ private PluginScanResult scanPluginPath(
...
             ClassLoader loader
     ) throws InstantiationException, IllegalAccessException {
-        Set<Class<? extends T>> plugins = reflections.getSubTypesOf(klass);
-
+        Set<Class<? extends T>> plugins;
+        try {
+             plugins = reflections.getSubTypesOf(klass);
+        } catch (ReflectionsException e) {
+            log.debug("Reflections scanner could not find any classes for URLs: " +
+                reflections.getConfiguration().getUrls(), e);
```
Comment (kkonstantine) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java]: a tab is missing here for alignment too.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java [hunk 9]
```diff
@@ -339,8 +340,15 @@ private PluginScanResult scanPluginPath(
...
+            return Collections.emptyList();
+        }
+
```
Comment (kkonstantine) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java]: nit: probably checkstyle won't catch this, but if you format the block, there's an extra tab that you can remove here too.

**Review by kkonstantine — COMMENTED** [review]
@ncliang this looks good overall, but there are a few checkstyle issues that fail the build.

**Comment by kkonstantine** [comment]
retest this please

**Comment by kkonstantine** [comment]
Rebased to get the recent fix on the unused imports.

**Comment by kkonstantine** [comment]
retest this please

**Review by kkonstantine — APPROVED** [review]
Thanks @ncliang 
This is a solid fix. LGTM!
I'll merge to `trunk` and `2.5`

**Comment by kkonstantine** [comment]
Merged to trunk and cherry-picked on 2.5
